### PR TITLE
[pulsar-broker] Support caching to drain backlog consumers

### LINKED ISSUE #12257: Slow backlog draining with high fanout topics
### Issue

When topics with high publish rate and high fan-out with large number of subscriptions or large number of replicators, build backlog then it's really difficult to drain the backlog for such topics while they serving high publish-rate. It's reproducible with multiple topics (100) with each 10K writes, 6 backlog replicators and multiple subscriptions. It becomes impossible to drain backlogs even if most of the cursors are draining similar backlog because a large number of cold-reads from bookie makes overall backlog draining slower and even bookie-cache are not able to keep up with cached entries and eventually had to do cold-reads. 

In such scenario, we would like to force broker to cache backlog reads so, other subscriotions/replicators draining similar backlog can drain faster without going to a bookie and reduce number of cold-reads from bookie. 
Note: Bookie has read-cache but it's not helping in this usecase because of backlog draining of multiple topics across multiple brokers. Bookie eventually returns `TooManyRequestsException` with high reads which also slows down the backlog draining. and we are not able to drain the backlog and topics with high publish rates are keep building the backlog.

Below is the stats of the topic which can be used to reproduce the issue where due to slow reads across many topics builds backlog on all topics and it's not possible to drain such backlog while incoming high msgRateIn.

```
msgRateIn : 6K, msgRateOut: 4.5K, local-backlog: 3M, msgRateOut per replicator region: 4K, replication backlog:4M

"replication" : {
    "us1" : {
      "msgRateIn" : 4637.9,
      "msgRateOut" : 4534.8,
      "msgRateExpired" : 0.0,
      "replicationBacklog" : 3976205,
      :
    "us2" : {
      "msgRateIn" : 4347.5,
      "msgRateOut" : 4439.6,
      "msgRateExpired" : 0.0,
      "replicationBacklog" : 3934344,
      :
	
    "us3" : {
      "msgRateIn" : 4254.2,
      "msgRateOut" : 4534.0,
      "msgRateExpired" : 0.0,
      "replicationBacklog" : 3974243,
      :
	  :
	  :
	  :
    "us6" : {
      "msgRateIn" : 4845.3,
      "msgRateOut" : 4485.4,
      "msgRateExpired" : 0.0,
      "replicationBacklog" : 3974529,
      :
```

### PR DESCRIPTION
**Pull request by rdhabalia**
### Motivation

It fixes #12257 

When topics with high publish rate and high fan-out with large number of subscriptions or large number of replicators, build backlog then it's really difficult to drain the backlog for such topics while they serving high publish-rate. It's reproducible with multiple topics (100) with each 10K writes, 6 backlog replicators and multiple subscriptions. It becomes impossible to drain backlogs even if most of the cursors are draining similar backlog because a large number of cold-reads from bookie makes overall backlog draining slower and even bookie-cache are not able to keep up with cached entries and eventually had to do cold-reads. 

In such scenario, we would like to force broker to cache backlog reads so, other subscriotions/replicators draining similar backlog can drain faster without going to a bookie and reduce number of cold-reads from bookie. 
Note: Bookie has read-cache but it's not helping in this usecase because of backlog draining of multiple topics across multiple brokers. Bookie eventually returns `TooManyRequestsException` with high reads which also slows down the backlog draining. and we are not able to drain the backlog and topics with high publish rates are keep building the backlog.

Below is the stats of the topic which can be used to reproduce the issue where due to slow reads across many topics builds backlog on all topics and it's not possible to drain such backlog while incoming high msgRateIn.

```
msgRateIn : 6K, msgRateOut: 4.5K, local-backlog: 3M, msgRateOut per replicator region: 4K, replication backlog:4M

"replication" : {
    "us1" : {
      "msgRateIn" : 4637.9,
      "msgRateOut" : 4534.8,
      "msgRateExpired" : 0.0,
      "replicationBacklog" : 3976205,
      :
    "us2" : {
      "msgRateIn" : 4347.5,
      "msgRateOut" : 4439.6,
      "msgRateExpired" : 0.0,
      "replicationBacklog" : 3934344,
      :
	
    "us3" : {
      "msgRateIn" : 4254.2,
      "msgRateOut" : 4534.0,
      "msgRateExpired" : 0.0,
      "replicationBacklog" : 3974243,
      :
	  :
	  :
	  :
    "us6" : {
      "msgRateIn" : 4845.3,
      "msgRateOut" : 4485.4,
      "msgRateExpired" : 0.0,
      "replicationBacklog" : 3974529,
      :
```

### Modification
- Add mechanism to allow broker to cache backlog reads
- By default this feature will be disabled and will not impact any existing workflow
- This feature can be tuned based on parameters which cache backlog reads if x-number of cursors are reading the same backlog to leverage cache (`managedLedgerMinimumBacklogCursorsForCaching`, `managedLedgerMaxBacklogBetweenCursorsForCaching`)

### Result 
Suppert broker cache to drain backlogs for high fanout topics.

### CODE REVIEW

**Comment by rdhabalia** [comment]
ping.

**Comment by rdhabalia** [comment]
@eolivelli why did we change milestone from 2.9 to 2.10 for this PR.
I really don't understand there are many PRs which are necessary to address real issue for brokers are moving out from the main release..

**Comment by rdhabalia** [comment]
can we merge this PR for the current master so, we don't forget the change for the next release.

**Comment by rdhabalia** [comment]
ping. can we merge this PR as it keep needs rebasing the PR.

**Comment by codelipenghui** [comment]
@rdhabalia Sorry for the late response, Since we have many unclosed PRs it's not easy to maintain, but we are trying to improve this situation see https://lists.apache.org/thread/0lm9tyjqtgtvwkfowkfhbxy24nh8tyxh, which can help us to focus on the active PRs and issues first.

It's a nice enhancement for high fanout read throughput, since we introduced new configurations, new caches, could you please help create a [proposal](https://github.com/apache/pulsar/wiki/Pulsar-Improvement-Proposal-%28PIP%29) and start to discuss it in the dev email thread? 

I plan to push the 2.10 release vote at the end of the month, I think we have time to include this enhancement. Of course, if you don't have time to deal with it right now, I think we can move it to 2.11?

**Comment by eolivelli** [comment]
@rdhabalia I am really sorry. I missed this PR.

Probablt I moved it when I prepared 2.9 and I did a bulk move of every uncommitted PRs.

I will review ASAP.

**Review by eolivelli — CHANGES_REQUESTED** [review]
Overall I am supportive of this proposal.
I left some minor comments.

Advertising about this on the mailing list is a good thing, because otherwise such important improvements may fall underestimated or totally missed because the flood of github notifications is huge

**Comment by rdhabalia** [comment]
this PR was created 6 months back. and thank you for blocking and taking this PR out of 2.10 release. I think if we want to unnecessarily block the PRs and not let them merge even though it's causing production issues then I don't appreciate such practice.

**Comment by rdhabalia** [comment]
I need to keep rebasing the PR due to unnecessary blocking. will rebase it.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 1]
```diff
@@ -2327,6 +2327,15 @@ public void checkBackloggedCursors() {
         });
     }
 
+    @Override
+    public void checkCursorsToCacheEntries() {
+        try {
+            ledger.checkCursorsToCacheEntries();
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (rdhabalia) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2344]: it doesn't throw `InterruptedException`.

[the remark this answers is not in the document]
Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2344]: When you catch generic Exception you should handle every possible cases and InterruptedException is always very annoying.

**Review by eolivelli — APPROVED** [review]
I don't want to block this patch.

@dave2wave you will be interested in testing performances of backlog draining once we merge this patch to master branch

**Comment by eolivelli** [comment]
I believe that this proposal is very valuable. I am working on problems related to caching, see https://github.com/apache/pulsar/issues/16421 

@rdhabalia would you mind rebase and starting the discussion on dev@ ?

**Comment by rdhabalia** [comment]
ok, I will rebase it soon.

**Review by dave2wave — APPROVED** [review]
It was a mistake to delay these improvements. These should provide enhanced consumption in many scenarios. et's make a priority to get this into 2.11!

**Comment by codelipenghui** [comment]
We should have a proposal for this change? The PR introduced a new feature and new configurations to the broker.

**Comment by dave2wave** [comment]
I ran the OMB on master after this merge and it worked well. I noticed no performance issues. This was with failover subscriptions. Someone should test shared subscriptions.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java [hunk 2]
```diff
@@ -213,8 +213,11 @@ public void safeRun() {
             EntryImpl entry = EntryImpl.create(ledger.getId(), entryId, data);
             // EntryCache.insert: duplicates entry by allocating new entry and data. so, recycle entry after calling
             // insert
-            ml.entryCache.insert(entry);
-            entry.release();
+            // Entry cache doesn't copy the data if entry already exist into the cache.
+            // Backlog read tries to add entry into cache which can try to add duplicate entry into cache.
+            if (ml.entryCache.insert(entry)) {
+                entry.release();
+            }
```
Comment (codelipenghui) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java:220]: Here will be a potential memory leak? If the cache doesn't have enough space, will the entry not be released?

Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java:220]: I think @codelipenghui is right that this is a memory leak. I think that this part of the change should be reverted.

Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpAddEntry.java:220]: I created #17007 to revert this.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 3]
```diff
@@ -4144,4 +4156,41 @@ public void checkInactiveLedgerAndRollOver() {
...
+                if (backlog < minBacklogEntriesForCaching) {
+                    continue;
+                }
+                if (Math.abs(backlog - other.getValue()) <= maxBacklogBetweenCursorsForCaching) {
+                    cursorsInSameBacklogRange++;
+                }
+            }
+            cursor.getKey().setCacheReadEntry(cursorsInSameBacklogRange >= minBacklogCursorsForCaching);
+            if (log.isDebugEnabled()) {
+                log.info("{} Enabling cache read = {} for {}", name,
```
Comment (Technoboy-) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:4191]: Oh, a mistake here.

File: conf/standalone.conf [hunk 4]
```diff
@@ -752,6 +752,16 @@ managedLedgerAddEntryTimeoutSeconds=0
...
 
+# Minimum cursors that must be in backlog state to cache and reuse the read entries.
+# (Default =0 to disable backlog reach cache)
+managedLedgerMinimumBacklogCursorsForCaching=0
+
+# Minimum backlog entries for any cursor before start caching reads.
+managedLedgerMinimumBacklogEntriesForCaching=100
+
+# Maximum backlog entry difference to prevent caching entries that can't be reused.
+managedLedgerMaxBacklogBetweenCursorsForCaching=10000
```
Comment (michaeljmarshall) [inline conf/standalone.conf:763]: @rdhabalia - was it intentional for these two settings to differ from the settings in the `ManagedLedgerConfig` class?

**Pull request merged**